portal-bridge/src/stats.rs

```diff
@@ -0,0 +1,329 @@
+            );
+            trace!(
+                "GossipReport: block#{}: epoch_accumulator - accepted {:?}",
+                self.block_number,
+                stats.accepted
+            );
+            trace!(
+                "GossipReport: block#{}: epoch_accumulator - transferred {:?}",
+                self.block_number,
+                stats.transferred
+            );
+        }
+    }
+
+    fn update(&mut self, content_key: HistoryContentKey, results: ContentStats) {
+        match content_key {
+            HistoryContentKey::BlockHeaderWithProof(_) => {
+                self.header_with_proof = Some(results);
+            }
+            HistoryContentKey::BlockBody(_) => {
+                self.block_body = Some(results);
+            }
+            HistoryContentKey::BlockReceipts(_) => {
+                self.receipts = Some(results);
+            }
+            HistoryContentKey::EpochAccumulator(_) => {
+                self.epoch_accumulator = Some(results);
+            }
+        }
+    }
+}
+
+// Struct to record the gossip stats for a single piece of content (eg key/value pair),
+// consolidating results from jsonrpc requests to 1/many clients.
+#[derive(Debug, Clone, Default)]
+pub struct ContentStats {
+    pub offered: Vec<Enr>,
+    pub accepted: Vec<Enr>,
+    pub transferred: Vec<Enr>,
+    pub failures: u64,
+}
+
+impl ContentStats {
+    pub fn report(&self) -> String {
+        format!(
+            "offered: {}, accepted: {}, transferred: {}, failures: {}",
+            self.offered.len(),
+            self.accepted.len(),
+            self.transferred.len(),
+            self.failures,
+        )
+    }
+}
+
+impl From<Vec<Result<TraceGossipInfo, Error>>> for ContentStats {
+    fn from(results: Vec<Result<TraceGossipInfo, Error>>) -> Self {
+        let mut content_stats = ContentStats::default();
+        for trace_gossip_info in results.iter() {
+            match trace_gossip_info {
+                Ok(info) => {
+                    for enr in info.offered.iter() {
+                        let enr = Enr::from_str(enr).expect("");
+                        // don't double count an enr if multiple clients offered the same content
+                        // to a peer
+                        if !content_stats.offered.contains(&enr) {
+                            content_stats.offered.push(enr);
+                        }
+                    }
+
+                    for enr in info.accepted.iter() {
+                        let enr = Enr::from_str(enr).expect("");
+                        if !content_stats.accepted.contains(&enr) {
+                            content_stats.accepted.push(enr);
+                        }
+                    }
+
+                    for enr in info.transferred.iter() {
+                        let enr = Enr::from_str(enr).expect("");
+                        if !content_stats.transferred.contains(&enr) {
+                            content_stats.transferred.push(enr);
```

Comment:
These ``expect("")`` should never fail. But currently they are being used as if they were ``unwrap()``. It would be nice to put a message saying ``x is impossible`` instead of an empty string. Just a nitpick so not a big deal if you don't want to, empty expect() just taste wrong idk.